packages/language-server/src/plugins/typescript/DocumentSnapshot.ts

```diff
@@ -1,24 +1,216 @@
+        let parserError: ParserError | null = null;
+        const scriptInfo = extractTag(text, 'script');
+        const styleInfo = extractTag(text, 'style');
+        let nrPrependedLines = 0;
+
+        if (isSvelteFilePath(uri)) {
+            try {
+                const tsx = svelte2tsx(text);
+                text = tsx.code;
+                tsxMap = tsx.map;
+                if (tsxMap) {
+                    tsxMap.sources = [uri];
+
+                    const tsCheck = scriptInfo?.content.match(tsCheckRegex);
+                    if (tsCheck) {
+                        text = `//${tsCheck[2]}${ts.sys.newLine}` + text;
+                        nrPrependedLines = 1;
+                    }
+                }
+            } catch (e) {
+                // Error start/end logic is different and has different offsets for line, so we need to convert that
+                const start: Position = {
+                    line: e.start?.line - 1 ?? 0,
+                    character: e.start?.column ?? 0,
+                };
+                const end: Position = e.end
+                    ? { line: e.end.line - 1, character: e.end.column }
+                    : start;
+                parserError = {
+                    range: { start, end },
+                    message: e.message,
+                    code: -1,
+                };
+                // fall back to extracted script, if any
+                text = scriptInfo ? scriptInfo.content : '';
+            }
+        }
+
+        return { tsxMap, text, scriptInfo, styleInfo, parserError, nrPrependedLines };
+    }
+
+    private constructor(
+        public version: number,
+        public readonly scriptKind: ts.ScriptKind,
+        public readonly filePath: string,
+        public readonly parserError: ParserError | null,
+        public readonly scriptInfo: TagInformation | null,
+        public readonly styleInfo: TagInformation | null,
+        private readonly text: string,
+        private readonly originalText: string,
+        private readonly nrPrependedLines: number,
+        private readonly tsxMap?: RawSourceMap,
+    ) {}
+
+    getText(start: number, end: number) {
+        return this.text.substring(start, end);
+    }
+
+    getLength() {
+        return this.text.length;
+    }
+
+    getChangeRange() {
+        return undefined;
+    }
+
+    positionAt(offset: number) {
+        return positionAt(offset, this.text);
+    }
+
+    async getFragment() {
+        if (!this.fragment) {
+            const uri = pathToUrl(this.filePath);
+            const mapper = !this.scriptInfo
+                ? new IdentityMapper()
+                : !this.tsxMap
+                ? new FragmentMapper(this.originalText, this.scriptInfo)
+                : new ConsumerDocumentMapper(
+                      await new SourceMapConsumer(this.tsxMap),
+                      uri,
+                      this.nrPrependedLines,
+                  );
```

Comment:
this is a lot happening on a single express, might be worth a refactor later

Reply:
Absolutely true, one of the first things I'm gonna look into after this is merged.